docstore/query.go

```diff
@@ -196,20 +197,100 @@ func (q *Query) Delete(ctx context.Context) error {
 	if err := q.validateWrite("delete"); err != nil {
 		return err
 	}
-	return q.coll.driver.RunDeleteQuery(ctx, q.dq)
+	type rdq interface {
```

Comment:
This interface should be defined in the driver package.

Reply:
@shantuo done.

@polinasok It's a common Go idiom that lets you dynamically ask the question, does some value implement an interface? It lets you override behavior in a general way. Some examples:
- https://godoc.org/encoding/json#Marshal, from "If an encountered value...".
- https://godoc.org/io#Copy, from "If dst implements..."

Reply:
Sorry, I didn't fully answer this. I told you in general about it but not why I'm using it here.

We want the mongo driver to do its optimized thing for these kinds of queries, but we want to default to a more general algorithm. This trick lets us do that without hard-coding mongo into the portable part of the code.